### PR TITLE
♻️ Refacto Rpc Methods

### PR DESCRIPTION
**Pull request by Tbelleng**
# Pull Request type

Please add the labels corresponding to the type of changes your PR introduces:

- Refacto
- Upgrade

## What is the current behavior?

♻️ All RPC methods are now available through a unique folder containing Read, Write and Trace methods.
🗑️ I also removed rpc-core crate and merged file into lib.rs and utils.rs (in rpc folder).
🆙 In addition, i changed some out-dated types from madara primitive and added starknet-rs types.

🚀 Next PR will aim to implement a routing through those new directories in order to run one Deoxys instance but letting the possibility to the user to choose the RPC version of his choice.

Resolves: #NA

## Does this introduce a breaking change?

No

### CODE REVIEW

File: crates/client/rpc/src/methods/trace/simulate_transactions.rs [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+            Err(_) => {
+                return Err(ConvertCallInfoToExecuteInvocationError::TransactionExecutionFailed);
+            }
+        }
+    }
+
+    Ok(results)
+}
+
+// fn try_get_function_invocation_from_call_info(
```
Comment (Trantorian1) [inline crates/client/rpc/src/methods/trace/simulate_transactions.rs]: Why is this commented code here? Maybe consider removing it if it non longer serves a purpose.

File: crates/client/rpc/src/methods/get_block.rs [hunk 2]
```diff
@@ -19,10 +17,10 @@ use starknet_core::types::{
...
 use crate::utils::{
-    l1_gas_price, new_root, parent_hash, sequencer_address, starknet_version, status, timestamp, tx_conv,
-    tx_hash_compute, tx_hash_retrieve,
+    get_block_by_block_hash, l1_gas_price, new_root, parent_hash, sequencer_address, starknet_version, status,
+    timestamp, tx_conv, tx_hash_compute, tx_hash_retrieve,
 };
-use crate::Starknet;
+use crate::{Felt, Starknet};
 
 pub(crate) fn get_block_with_tx_hashes_finalized<A, B, BE, G, C, P, H>(
```
Comment (Trantorian1) [inline crates/client/rpc/src/methods/get_block.rs]: Consider refactoring this into the newer `get_block_with_tx_hashes.rs` in `read/` to keep the code more organised.

File: crates/client/rpc/src/utils.rs [hunk 3]
```diff
@@ -165,3 +185,189 @@ pub(crate) fn l1_gas_price(block: &mp_block::Block) -> ResourcePrice {
...
+pub fn casm_entry_point_to_compiled_entry_point(value: &CasmContractEntryPoint) -> CompiledClassEntrypoint {
+    CompiledClassEntrypoint {
+        selector: biguint_to_field_element(&value.selector),
+        offset: value.offset.try_into().unwrap(),
+        builtins: value.builtins.clone(),
+    }
+}
+
+/// Converts a [BigUint] to a [FieldElement]
+pub fn biguint_to_field_element(value: &BigUint) -> FieldElement {
```
Comment (Trantorian1) [inline crates/client/rpc/src/utils.rs]: This should be private (only used in `utils`)

File: crates/client/rpc/src/utils.rs [hunk 4]
```diff
@@ -165,3 +185,189 @@ pub(crate) fn l1_gas_price(block: &mp_block::Block) -> ResourcePrice {
...
+pub fn casm_entry_points_to_compiled_entry_points(value: &CasmContractEntryPoints) -> CompiledClassEntrypointList {
+    CompiledClassEntrypointList {
+        external: value.external.iter().map(casm_entry_point_to_compiled_entry_point).collect(),
+        l1_handler: value.l1_handler.iter().map(casm_entry_point_to_compiled_entry_point).collect(),
+        constructor: value.constructor.iter().map(casm_entry_point_to_compiled_entry_point).collect(),
+    }
+}
+
+/// Converts a [CasmContractEntryPoint] to a [CompiledClassEntrypoint]
+pub fn casm_entry_point_to_compiled_entry_point(value: &CasmContractEntryPoint) -> CompiledClassEntrypoint {
```
Comment (Trantorian1) [inline crates/client/rpc/src/utils.rs]: This should be private (only used in `utils`)

File: crates/client/rpc/src/utils.rs [hunk 5]
```diff
@@ -165,3 +185,189 @@ pub(crate) fn l1_gas_price(block: &mp_block::Block) -> ResourcePrice {
...
+        compiler_version: casm_contract_class.compiler_version.clone(),
+        bytecode: casm_contract_class.bytecode.iter().map(|x| biguint_to_field_element(&x.value)).collect(),
+        entry_points_by_type: casm_entry_points_to_compiled_entry_points(&casm_contract_class.entry_points_by_type),
+        hints: vec![],        // not needed to get class hash so ignoring this
+        pythonic_hints: None, // not needed to get class hash so ignoring this
+    }
+}
+
+/// Converts a [CasmContractEntryPoints] to a [CompiledClassEntrypointList]
+pub fn casm_entry_points_to_compiled_entry_points(value: &CasmContractEntryPoints) -> CompiledClassEntrypointList {
```
Comment (Trantorian1) [inline crates/client/rpc/src/utils.rs]: This should be private (only used in `utils`)

File: crates/client/rpc/src/utils.rs [hunk 6]
```diff
@@ -165,3 +185,189 @@ pub(crate) fn l1_gas_price(block: &mp_block::Block) -> ResourcePrice {
...
+}
+
+// Utils to convert Casm contract class to Compiled class
+pub fn get_casm_cotract_class_hash(casm_contract_class: &CasmContractClass) -> FieldElement {
+    let compiled_class = casm_contract_class_to_compiled_class(casm_contract_class);
+    compiled_class.class_hash().unwrap()
+}
+
+/// Converts a [CasmContractClass] to a [CompiledClass]
+pub fn casm_contract_class_to_compiled_class(casm_contract_class: &CasmContractClass) -> CompiledClass {
```
Comment (Trantorian1) [inline crates/client/rpc/src/utils.rs]: This should be private (only used in `utils`)

File: crates/client/rpc/src/utils.rs [hunk 7]
```diff
@@ -165,3 +185,189 @@ pub(crate) fn l1_gas_price(block: &mp_block::Block) -> ResourcePrice {
...
+
+    let constructor = collect_entry_points(entries, EntryPointType::Constructor)?;
+    let external = collect_entry_points(entries, EntryPointType::External)?;
+    let l1_handler = collect_entry_points(entries, EntryPointType::L1Handler)?;
+
+    Ok(LegacyEntryPointsByType { constructor, external, l1_handler })
+}
+
+/// Returns a [LegacyContractEntryPoint] (starknet-rs) from a [EntryPoint] (starknet-api)
+pub fn to_legacy_entry_point(entry_point: EntryPoint) -> Result<LegacyContractEntryPoint, FromByteArrayError> {
```
Comment (Trantorian1) [inline crates/client/rpc/src/utils.rs]: This should be private (only used in `utils`)

File: crates/client/rpc/src/utils.rs [hunk 8]
```diff
@@ -165,3 +185,189 @@ pub(crate) fn l1_gas_price(block: &mp_block::Block) -> ResourcePrice {
...
+    // https://github.com/keep-starknet-strange/blockifier/blob/no_std-support-7578442/crates/blockifier/src/execution/contract_class.rs#L129
+    // https://github.com/keep-starknet-strange/blockifier/blob/no_std-support-7578442/crates/blockifier/src/execution/contract_class.rs#L389
+    // serde_json::to_writer(&mut gzip_encoder, data)?;
+    gzip_encoder.write_all(data)?;
+    Ok(gzip_encoder.finish()?)
+}
+
+/// Returns a [Result<LegacyEntryPointsByType>] (starknet-rs type) from a [HashMap<EntryPointType,
+/// Vec<EntryPoint>>]
+pub fn to_legacy_entry_points_by_type(
```
Comment (Trantorian1) [inline crates/client/rpc/src/utils.rs]: This should be private (only used in `utils`)

File: crates/client/rpc/src/utils.rs [hunk 9]
```diff
@@ -165,3 +185,189 @@ pub(crate) fn l1_gas_price(block: &mp_block::Block) -> ResourcePrice {
...
+        storage_diffs,
+        deprecated_declared_classes,
+        declared_classes,
+        deployed_contracts,
+        replaced_classes,
+    }
+}
+
+/// Returns a compressed vector of bytes
+pub(crate) fn compress(data: &[u8]) -> Result<Vec<u8>> {
```
Comment (Trantorian1) [inline crates/client/rpc/src/utils.rs]: This should be private (only used in `utils`)

File: crates/client/rpc/src/methods/read/get_state_update.rs [hunk 10]
```diff
@@ -15,9 +14,10 @@ use sp_api::ProvideRuntimeApi;
...
 use sp_runtime::traits::Block as BlockT;
 use starknet_api::block::BlockHash;
-use starknet_core::types::{FieldElement, MaybePendingStateUpdate, StateDiff, StateUpdate};
+use starknet_core::types::{BlockId, BlockTag, FieldElement, MaybePendingStateUpdate, StateDiff, StateUpdate};
 
 use crate::errors::StarknetRpcApiError;
+use crate::utils::get_block_by_block_hash;
 use crate::Starknet;
 
 pub(crate) fn get_state_update_finalized<A, B, BE, G, C, P, H>(
```
Comment (Trantorian1) [inline crates/client/rpc/src/methods/read/get_state_update.rs]: This should be private

File: crates/client/rpc/src/methods/read/get_transaction_receipt.rs [hunk 11]
```diff
@@ -29,9 +28,9 @@ use starknet_core::types::{
 use crate::errors::StarknetRpcApiError;
 use crate::utils::{
     blockifier_call_info_to_starknet_resources, extract_events_from_call_info, extract_messages_from_call_info,
-    tx_hash_compute, tx_hash_retrieve,
+    get_block_by_block_hash, tx_hash_compute, tx_hash_retrieve,
 };
-use crate::Starknet;
+use crate::{Felt, Starknet, StarknetReadRpcApiServer};
 
 pub(crate) fn get_transaction_receipt_finalized<A, B, BE, G, C, P, H>(
```
Comment (Trantorian1) [inline crates/client/rpc/src/methods/read/get_transaction_receipt.rs]: This should be private

Comment (Trantorian1) [reply crates/client/rpc/src/methods/read/get_transaction_receipt.rs]: same for `get_transaction_receipt_pending`

File: crates/client/rpc/src/methods/read/get_state_update.rs (hunk 10 above)
Comment (Trantorian1) [reply crates/client/rpc/src/methods/read/get_state_update.rs]: Same for `get_state_update_pending`

File: crates/client/rpc/src/methods/read/lib.rs [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+use super::get_storage_at::*;
+use super::get_transaction_by_block_id_and_index::*;
+use super::get_transaction_by_hash::*;
+use super::get_transaction_receipt::*;
+use super::get_transaction_status::*;
+use super::syncing::*;
+use crate::{Felt, Starknet, StarknetReadRpcApiServer};
+
+#[async_trait]
+#[allow(unused_variables)]
```
Comment (Trantorian1) [inline crates/client/rpc/src/methods/read/lib.rs]: Why is this here?

File: crates/client/rpc/src/methods/read/syncing.rs [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+///
+/// ### Returns
+///
+/// * `Syncing` - An Enum that can either be a `mc_rpc_core::SyncStatus` struct representing the
+///   sync status, or a `Boolean` (`false`) indicating that the node is not currently synchronizing.
+///
+/// This is an asynchronous function due to its reliance on `sync_service.best_seen_block()`,
+/// which potentially involves network communication and processing to determine the best block
+/// seen by the sync service.
+#[allow(unused_variables)]
```
Comment (Trantorian1) [inline crates/client/rpc/src/methods/read/syncing.rs]: Same, why are we ignoring unused variables?

Comment (Trantorian1) [reply crates/client/rpc/src/methods/read/syncing.rs]: This appears to be the case for every RPC method. Please remove this as it risks obfuscating future errors.

File: crates/client/rpc/src/methods/read/tools.rs [hunk 14]
```diff
@@ -0,0 +1,27 @@
+use std::sync::Arc;
+
+use pallet_starknet_runtime_api::{ConvertTransactionRuntimeApi, StarknetRuntimeApi};
+use sp_api::ProvideRuntimeApi;
+use sp_runtime::traits::Block as BlockT;
+use sp_runtime::DispatchError;
+
+use crate::errors::StarknetRpcApiError;
+
+pub fn convert_error<C, B, T>(
```
Comment (Trantorian1) [inline crates/client/rpc/src/methods/read/tools.rs]: Why is this not in `utils`?

File: crates/client/rpc/src/methods/trace/trace_implementation.rs [hunk 15]
```diff
@@ -25,19 +21,21 @@ use sp_blockchain::HeaderBackend;
...
+use super::lib::*;
 use crate::errors::StarknetRpcApiError;
-use crate::Starknet;
+use crate::utils::get_block_by_block_hash;
+use crate::{Starknet, StarknetReadRpcApiServer};
 
-#[async_trait]
 #[allow(unused_variables)]
-impl<A, B, BE, G, C, P, H> StarknetTraceRpcApiServer for Starknet<A, B, BE, G, C, P, H>
+pub async fn trace_block_transactions<A, B, BE, G, C, P, H>(
```
Comment (Trantorian1) [inline crates/client/rpc/src/methods/trace/trace_implementation.rs]: Perhaps use a different file for `trace_block_transactions` and `trance_transactions`, following the same pattern as the rest of the refactoring.

**Review by Trantorian1 — CHANGES_REQUESTED** [review]
Please fix the issues mentioned.

File: crates/client/rpc/src/methods/trace/trace_implementation.rs (hunk 15 above)
Comment (Tbelleng) [reply crates/client/rpc/src/methods/trace/trace_implementation.rs]: Yes but i implemented those methods in madara and separate these methods is not really necessary bc they use exactly the same pattern so its not very relevant to separate them according to me (like the difference between these 2 is only a empty vector)

**Review by antiyro — APPROVED** [review]

**Review by Trantorian1 — APPROVED** [review]
